### DISCUSSION TITLE
Help adding support for Nix lang needed

### DISCUSSION THREAD

**Discussion by pogopaule** [Wansmer/treesj#70] [Q&A] [answered]
I try to add support for [Nix](https://nixos.wiki/wiki/Overview_of_the_Nix_Language).

I am not sure if treesj in its current form can handle some "special" Nix formattings and [tree-sitter-nix](https://github.com/cstrahan/tree-sitter-nix) specifics or if I just don't understand treesj good enough to express these things.

The easiest way to get what I mean is to check out my [nix branch](https://github.com/pogopaule/treesj/tree/feat/nix) and run the tests. (See also [this diff](https://github.com/Wansmer/treesj/compare/main...pogopaule:treesj:feat/nix) between main and feat/nix)

## Problem 1

What I struggle with the most is that tree-sitter-nix parses an attribute set

```nix
{
  a = 1;
  b = 2;
  c = "c";
}
```

into 

```
element: attrset_expression [20, 4] - [24, 5]
  binding_set [21, 6] - [23, 14]
    binding: binding [21, 6] - [21, 12]
      attrpath: attrpath [21, 6] - [21, 7]
        attr: identifier [21, 6] - [21, 7]
      expression: integer_expression [21, 10] - [21, 11]
    binding: binding [22, 6] - [22, 12]
      attrpath: attrpath [22, 6] - [22, 7]
        attr: identifier [22, 6] - [22, 7]
      expression: integer_expression [22, 10] - [22, 11]
    binding: binding [23, 6] - [23, 14]
      attrpath: attrpath [23, 6] - [23, 7]
        attr: identifier [23, 6] - [23, 7]
      expression: string_expression [23, 10] - [23, 13]
        string_fragment [23, 11] - [23, 12]
```

With my current knowledge of treesj I am only able to split an attribute set `{ a = 1; b = 2; c = "c"; }` into either

```nix
{  a = 1;
  b = 2;
  c = "c"; }
```
with the preset `binding_set = ...`

or

```nix
{
  a = 1;  b = 2; c = "c";
}
```
with the preset `attrset_expression`.

I tried to work with `recursive = true` in various forms but did not succeed.

## Problem 2

The other problem I struggle with is that Nix has a very unusual formatting when splitting "function arguments". Take a look at the placement of the commas in https://github.com/pogopaule/treesj/blob/446df8a2833bdda43817dbb78302ce480946225e/tests/sample/index.nix#L29-L41 

## Potential workarounds

In both cases I can make things work by using treesj and the Nix formatter in combination. So for example I split an attribute set into
```nix
{  a = 1;
  b = 2;
  c = "c"; }
```
and let the formatter do the rest of the work. Same applies for the strange comma placement for function arguments.

This would kind of work but I am not satisfied with it.

**Comment by Wansmer** [comment, marked as answer]
Hello!

## Problem 1:
```lua
{
  attrset_expression = {
    target_nodes = { 'binding_set' }, -- because target node for format is not attrset_expression, but binding_set
  },
  binding_set = u.set_preset_for_dict({
    both = {
      non_bracket_node = true, -- It will created empty framing nodes for get new line after '{' and before '}'
      separator = ';',
      last_separator = true,
    },
  }),
}
```

## Problem 2:

Hmm, I see what all children of `formals` should be with same indents, but at the moment all inner indents calculating with `first indent + shiftwidth`.

Maximum what can be settings with TreeSJ it:
```nix
# from
  func1 =
    { foo, bar, blub, ... }:
    { };
# to
  func1 =
    { foo
      , bar
      , blub
      , ...
    }:
    { };
```

via

```lua
{
  formals = u.set_preset_for_args({
    both = {
      omit = { 'formal', 'ellipses' }
    },
    split = {
      separator = '',
    },
    join = {
      separator = ',',
      space_in_brackets = true,
    },
  }),
}
```

But this form adding extra space before separator, when joining. Looks like bug, will fix it.

## About `let_expression`:

Via
```lua
{
  let_expression = { target_nodes = { 'binding_set' } },
}
```

It can be transformed

```nix
  bla =
    let x=1; y=2; in x + y;

  blu =
    let
      x = 1;
      y = 2;
    in x + y;
```

At the moment, no way to move `x + y` at new line, unfortunately.

Thanks for sharing these questions. I'll take them into account in further development and improving TreeSJ.

**Reply by pogopaule** [reply]
Thank you very much for your quick and detailed answer! I think those solutions are good enough for the moment. They do 99% of what is important. The nix formatter will do the rest.

**Reply by Wansmer** [reply]
Btw, I fixed a problem with extra spaces (094a0df83ed6398b707f0b4460e3110f5ab76ba4) and added option for inner indent (120ec4473edb01bb1db1b73a984237bd63cc3d8e) – you can use it for `formals`.

**Comment by Wansmer** [comment]
Add preset for `let_expression` with #74
